python/dgl/contrib/kvstore/kvstore.py

```diff
@@ -0,0 +1,186 @@
+# This file contains DGL distributed kvstore APIs.
+from ...network import _create_sender, _create_receiver
+from ...network import _finalize_sender, _finalize_receiver
+from ...network import _recv_ip_port, _send_local_ip_port
+from ...network import _recv_graph_store_msg, _send_graph_store_msg
+
+import time
+import signal
+from enum import Enum
+from collections import namedtuple
+
+_WAIT_TIME_SEC = 3  # 3 seconds for socket sync
+
+def SocketSync():
+    time.sleep(_WAIT_TIME_SEC)
+
+def keyboardInterruptHandler(signal, frame):
+    """Users use [Ctl + C] to exit the KVStore service
+    """
+    print("KeyboardInterrupt (ID: {}) has been caught. Cleaning up DGL ...".format(signal))
+    exit(0)
+
+signal.signal(signal.SIGINT, keyboardInterruptHandler)
+
+KVStoreMsg = namedtuple("KVStoreMsg", "type feat_name node_ids node_feats")
+
+class MessageType(Enum):
+    """For now, we only support two types of message
+    """
+    PUSH = 0
+    PULL = 1
+
+class KVStore(object):
+    """KVStore is a key-value store for DGL distributed training.
+
+    In practice, we use KVStore to hold large-scale node features or node embedding 
+    in standalone machine with big memory capability. 
+
+    KVStore provides three APIs for KVStore-server:
+
+      * init_server(ip, port, node_feats) initialize KVStore-server.
+
+      * push_msg_handle(msg) user-defined handler for push message.
+
+      * pull_msg_handle(msg) user-defined handler for pull message.
+
+    KVStore also provides three APIs for KVStore-client:
+
+      * init_client(sever_ip, server_port) initialize KVStore-client.
+
+      * pull(feat_name, node_ids) requests the feature values associated with corresponding 
+        feature name and node IDs.
+
+      * push(feat_name, node_ids, node_feats) sends feature name, node IDs, and node features 
+        to the KVStore-server, which uses these data to updating graph embeddings.
```

Comment:
I think it's better to split the server and the client. It's quite confusing why these two components are in the same class.

Reply:
Ok, I will do this change.